src/ctx.rs

```diff
@@ -1,147 +1,240 @@
+/// ```
+/// # use rav1d::case_set;
+/// # use rav1d::src::disjoint_mut::DisjointMut;
+/// let mut buf = DisjointMut::new([0u8; 32]);
+/// let len = 16;
+/// for offset in [0, 16] {
+///     case_set!(32, len, offset, {
+///         set_disjoint!(&mut buf, 1u8);
+///     });
+/// }
+/// ```
+///
+/// ## Multiple buffer ranges
+///
+/// Multiple buffers with different lengths and offsets can be filled with the
+/// same body statements. In the following example, two buffers with different
+/// sizes are initialized by quarters.
+/// ```
+/// # use rav1d::case_set;
+/// let mut buf1 = [0u8; 32];
+/// let mut buf2 = [0u8; 64];
+/// for offset in [0, 8, 16, 24] {
+///     case_set!(16, buf=[&mut buf1[..], &mut buf2[..]], len=[8, 16], offset=[offset, offset*2], {
+///         set!(buf, len as u8 >> 3);
+///     });
+/// }
+/// ```
 ///
-/// For power of 2 lengths `<= UP_TO`,
-/// the `memset` is done as an array write of that exactly (compile-time known) length.
-/// If the length is not a power of 2 or `> UP_TO`,
-/// then the `memset` is done by [`slice::fill`] (a `memset` call) if `WITH_DEFAULT` is `true`,
-/// or else skipped if `WITH_DEFAULT` is `false`.
+/// A more realistic example of filling multiple buffers with the same value is
+/// initializing different struct fields at the same time (from
+/// `src/decode.rs`):
+/// ```ignore
+/// case_set!(32, ctx=[(&t.l, 1), (&f.a[t.a], 0)], len=[bh4, bw4], offset=[by4, bx4], {
+///      let (dir, dir_index) = ctx;
+///      set_disjoint!(dir.seg_pred, seg_pred.into());
+///      set_disjoint!(dir.skip_mode, b.skip_mode);
+///      set_disjoint!(dir.intra, 0);
+///      set_disjoint!(dir.skip, b.skip);
+///      set_disjoint!(dir.pal_sz, 0);
+/// });
+/// ```
 ///
-/// This optimizes for the common cases where `buf.len()` is a small power of 2,
-/// where the array write is optimized as few and large stores as possible.
-#[inline]
-pub fn small_memset<T: Clone + Copy, const UP_TO: usize, const WITH_DEFAULT: bool>(
-    buf: &mut [T],
-    val: T,
-) {
-    fn as_array<T: Clone + Copy, const N: usize>(buf: &mut [T]) -> &mut [T; N] {
-        buf.try_into().unwrap()
-    }
-    match buf.len() {
-        01 if UP_TO >= 01 => *as_array(buf) = [val; 01],
-        02 if UP_TO >= 02 => *as_array(buf) = [val; 02],
-        04 if UP_TO >= 04 => *as_array(buf) = [val; 04],
-        08 if UP_TO >= 08 => *as_array(buf) = [val; 08],
-        16 if UP_TO >= 16 => *as_array(buf) = [val; 16],
-        32 if UP_TO >= 32 => *as_array(buf) = [val; 32],
-        64 if UP_TO >= 64 => *as_array(buf) = [val; 64],
-        _ => {
-            if WITH_DEFAULT {
-                buf.fill(val)
+/// [`DisjointMut`]: crate::src::disjoint_mut::DisjointMut
+macro_rules! case_set {
+    ($UP_TO:literal, $(@DEFAULT=$WITH_DEFAULT:literal,)? $ctx:ident=[$($ctx_expr:expr),* $(,)?], $len:ident=[$($len_expr:expr),* $(,)?], $offset:ident=[$($offset_expr:expr),* $(,)?], $body:block) => {
```

Comment:
What's the `@` in `@DEFAULT` for?  And can we just name it `WITH_DEFAULT`?

Reply:
The `@` distinguishes it from `$ctx:ident` (`@` is not a valid identifier, or anything else in prefix position for that matter).

Reply:
What I really wanted was a way to pass in the boolean to unify the implementations but have two different macros for this (`case_set!` and `case_set_with_default!` or whatever).

Reply:
It might work if you put it at the end of the args?  Especially if `len` and `offset` are not caller named.  This way is okay, too, though, just funny looking I guess.